### ISSUE TITLE
Split module into `main` and `core`

### ISSUE THREAD

**Issue by philpl** [styled-components/styled-components#687]
We could split the bundle into a `core` module and the normal `main` one.

We can structure the library so that `main`, `native` and `no-parser` use `core`, so that the `no-parser` module doesn't conflict with `main` and people can use both alongside each other.

~Furthermore we can make main have the `domElements` and support `styled.div`, while the `styled` export on `core` could only support `styled('div')`. This way people can decide to make a really tiny bundle, if they need to.~

This would also solve the problem that libraries might not use `no-parser` and the user does, or vice versa.

/cc @mxstbr @geelen

**Comment by kentcdodds** [comment]
Was just about to file this issue myself 🎉

**Comment by philpl** [comment]
@kentcdodds I took the opportunity myself, since this could also solve the unrelated `no-parser` issue that we have.

**Comment by YasserKaddour** [comment]
Awesome 🎉, I proposed something like that [here](https://github.com/styled-components/styled-components/pull/640#issuecomment-291174320), I think this is far better than the hack I did for the `no-parser` entry.

 I could take a stab at this if no one mind, with a Lerna monoropo, once the `v2` is merged into `master` as discussed [here](https://github.com/styled-components/styled-components/issues/588#issuecomment-290907115).

I love the bytes war :crossed_swords: , it should be a tag.

**Comment by philpl** [comment]
@YasserKaddour I don't think it has to be a separate package. It could live in the main repo.

On a side note, once we have this we can add a transpilation step that transpiles styled.div and the likes to styled("div")

**Comment by YasserKaddour** [comment]
Sure,  but with separate packages every module could distribute an ES module.

**Comment by geelen** [comment]
Hmm, I don't mind breaking things apart and I certainly like smaller bundles, but I'd like to keep things grounded a bit. Splitting things in half where there's a natural division (like core/main) is good, and if that results in a smaller bundle then great, but I don't think we should be prioritising bundle size over DX, performance, or maintenance complexity. 11kb is fine for a lot of cases, the no-parser one will be even better, and anywhere we can improve weight without compromising the others then go for it. But just keep an eye on the broader goals of the project as you look for opportunities to split things up/slim things down.

That said, core vs main sounds like a good split. I _definitely_ wouldn't do the `styled("div")` thing though, but hoisting `styled.div` into `import { div } from 'styled/elements'` could make sense (and will be important for my current whitelisting stuff)

**Comment by philpl** [comment]
@geelen in that case let's do an elements file first, which should be simple, and a main/core with the goal of no-parser interoperability.

Sounds good? :)

**Comment by mxstbr** [comment]
Closing in favor of #588 since we're discussing this there.
